packages/opentelemetry-resource-detector-aws/test/detectors/AwsEcsDetector.test.ts

```diff
@@ -0,0 +1,203 @@
+ * limitations under the License.
+ */
+
+import * as assert from 'assert';
+import * as sinon from 'sinon';
+import {
+  awsEcsDetector,
+  AwsEcsDetector,
+} from '../../src/detectors/AwsEcsDetector';
+import {
+  assertEmptyResource,
+  assertContainerResource,
+} from '@opentelemetry/resources/test/util/resource-assertions';
+import { NoopLogger } from '@opentelemetry/core';
+import * as os from 'os';
+
+describe('BeanstalkResourceDetector', () => {
+  const errorMsg = {
+    fileNotFoundError: new Error('cannot find cgroup file'),
+  };
+
+  const correctCgroupData =
+    'abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklm';
+  const noisyCgroupData = `\n\n\n abcdefghijklmnopqrstuvwxyz \n ${correctCgroupData}`;
+  const multiValidCgroupData = `${correctCgroupData}\nabc${correctCgroupData}\nbcd${correctCgroupData}`;
+  const hostNameData = 'abcd.test.testing.com';
+
+  let readStub, hostStub;
+  let sandbox: sinon.SinonSandbox;
+
+  beforeEach(() => {
+    sandbox = sinon.createSandbox();
+    process.env.ECS_CONTAINER_METADATA_URI_V4 = '';
+    process.env.ECS_CONTAINER_METADATA_URI = '';
+  });
+
+  afterEach(() => {
+    sandbox.restore();
+  });
+
+  it('should successfully return resource data', async () => {
+    process.env.ECS_CONTAINER_METADATA_URI_V4 = 'ecs_metadata_v4_uri';
+    hostStub = sandbox.stub(os, 'hostname').returns(hostNameData);
+    readStub = sandbox
+      .stub(AwsEcsDetector, 'readFileAsync' as any)
+      .resolves(correctCgroupData);
+
+    const resource = await awsEcsDetector.detect({
+      logger: new NoopLogger(),
+    });
+
+    sandbox.assert.calledOnce(hostStub);
+    sandbox.assert.calledOnce(readStub);
+    assert.ok(resource);
+    assertContainerResource(resource, {
+      name: 'abcd.test.testing.com',
+      id: 'bcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklm',
+    });
+  });
+
+  it('should successfully return resource data with noisy cgroup file', async () => {
+    process.env.ECS_CONTAINER_METADATA_URI = 'ecs_metadata_v3_uri';
+    hostStub = sandbox.stub(os, 'hostname').returns(hostNameData);
+    readStub = sandbox
+      .stub(AwsEcsDetector, 'readFileAsync' as any)
+      .resolves(noisyCgroupData);
+
+    const resource = await awsEcsDetector.detect({
+      logger: new NoopLogger(),
+    });
+
+    sandbox.assert.calledOnce(hostStub);
+    sandbox.assert.calledOnce(readStub);
+    assert.ok(resource);
+    assertContainerResource(resource, {
+      name: 'abcd.test.testing.com',
+      id: 'bcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklm',
+    });
+  });
+
+  it('should always return first valid line of data', async () => {
+    process.env.ECS_CONTAINER_METADATA_URI = 'ecs_metadata_v3_uri';
+    hostStub = sandbox.stub(os, 'hostname').returns(hostNameData);
+    readStub = sandbox
+      .stub(AwsEcsDetector, 'readFileAsync' as any)
+      .resolves(multiValidCgroupData);
```

Comment:
I'm not sure this actually tests what it says.

each of the lines in `multiValidCgroupData` contain the same ID as the last 64 chars, so it doesn't necessarily prove it grabbed the first 'valid' (length > 64) one. it is minor, i can tell the code is working but perhaps a more accurate test would be to have the first line be 64 chars of a different/unexpected ID as well as having the last line be a valid, but unexpected ID to show you are grabbing the item that is expected (the middle one).

Reply:
Yes, in practice, multiple lines contain container ID in the same time. This testcase is to verify the logic here because we only grab the first valid container ID to reduce time cost.
I also think I can add an unexpected ID here. Thank you for pointing out!